### PR TITLE
Add dock-right shell, better questions ordering

### PR DESCRIPTION
**Pull request by sajacy**
Addresses #78 , pulls the shell to a "dock-right" layout on wide screens.

- [x] Change the order of the questions to group across categories of review to follow the actual thinking process that goes behind evaluating a manuscript from big picture to more detailed and/or less research related feedback (e.g., evaluation of specific parts of review, soundness of the science, etc., leaving the questions about open data at the end)

- [x] ~~Change the spacing between questions and font size to increase readability. Right there is too much space between questions and selection points for the answer.~~ With the compressed width of the dock-right UX, this didn't seem necessary.

- [x] ~~Change viz and spacing between selection points to select answer so that it is easier to see.~~  With the dock-right UX, this didn't seem necessary.

- [x] Allow for user to ~~drag~~ the review panel to the side of the preprint and organize the viz as they prefer and better fits their screen (ideal but not a priority). 

The UI is actually automatically restyled by window width: on `--wide-screen`-sized browser windows, dock-right layout is the only option, and resize controls are removed.  On mobile, the UX is unchanged (resizable-drawer layout).

Screenshots (on Mac)

Chrome
image

Safari (needs user intervention to zoom to "Actual Size")
image

Firefox
image

### CODE REVIEW

File: src/components/rapid-form-fragment.js [hunk 1]
```diff
@@ -13,26 +13,21 @@ export default function RapidFormFragment({ answerMap = {}, onChange }) {
...
   );
 
-  const freeFormQuestions = QUESTIONS.filter(({ type }) => type === 'Question');
-
   return (
     <div className="rapid-form-fragment">
-      <fieldset className="rapid-form-fragment__multi-choice-questions">
-        {yesNoQuestions.map(({ identifier, question, help, required }, i) => {
+      <fieldset className="rapid-form-fragment__text-response-questions">
+        {QUESTIONS.map(({ type, identifier, question, help, required }) => {
```
Comment (sajacy) [inline src/components/rapid-form-fragment.js:26]: The `QUESTIONS` variable is now rendered in-order.

File: src/components/shell.css [hunk 2]
```diff
@@ -58,66 +60,75 @@
...
+        min-width: 16px;
+        border: 2px solid var(--prereview-black);
+
+        /* width: 16px;
+        height: 16px; */
+      }
 
-    /* Add enough space so that finger fits on mobile (otherwise minimize and maximize are too close */
-    @media (--narrow-screen) {
-      &.shell__controls__button--maximize {
```
Comment (sajacy) [inline src/components/shell.css:120]: Removed an empty style declaration for `.shell__controls__button--maximize`

File: src/components/shell.css [hunk 3]
```diff
@@ -58,66 +60,75 @@
   height: var(--shell-header-height);
   box-sizing: border-box;
   position: fixed;
-  /* left: calc(50vw - var(--shell-control-icon-width) - var(--shell-control-icon-width) / 2); */
-  left: 10vw;
-  width: 80vw;
```
Comment (sajacy) [inline src/components/shell.css:63]: Moved to `@media (--wide-screen)` section, and modified to be `left:50vw, width:50vw`

File: src/components/shell.css [hunk 4]
```diff
@@ -58,66 +60,75 @@
...
-  @media (--narrow-screen) {
-    left: 0;
-    width: 100vw;
-  }
   & .shell__controls__left {
     display: flex;
     align-items: center;
     flex: 1 1 50%;
   }
-  & .shell__controls__right {
```
Comment (sajacy) [inline src/components/shell.css:79]: This diff makes this more confusing than it is: all the `.shell__controls*` are now wrapped in `@media (--narrow-screen)`

**Comment by sajacy** [comment]
@sballesteros @georgiamoon

**Comment by sajacy** [comment]
and tagging @dwins FTW?

File: src/utils/preprints.js [hunk 5]
```diff
@@ -54,7 +54,8 @@ export function getPdfUrl(preprint = {}) {
       encoding.contentUrl && encoding.encodingFormat === 'application/pdf'
   );
 
-  return encoding && encoding.contentUrl;
+  // doesn't seem to do anything, but FitH should set PDF view to Fit Horizontally
+  return encoding && (encoding.contentUrl+"?#view=FitH");
```
Comment (sajacy) [inline src/utils/preprints.js:58]: Adobe's "Open" parameter [docs](https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/pdf_open_parameters.pdf)

File: src/utils/render.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export function debounce(fn, ms) {
```
Comment (dwins) [inline src/utils/render.js:1]: We already have a dependency on lodash, can we just use their debounce function? https://lodash.com/docs/4.17.15#debounce

**Comment by dwins** [comment]
Thanks for contributing this improvement!  I left a couple of comments.  One higher level thought I had was that it might not always make sense to split the screen 50/50 in wide mode; I wonder if there's a better rule we could follow (cap the width of the form around 350px or so?)

These changes seem pretty safe so once you've responded to my comments I think we can approve this for merge.

File: src/utils/preprints.js (hunk 5 above)
Comment (sballesteros) [reply src/utils/preprints.js:58]: `contentUrl` could already have some query string parameters. I would vote to remove that.
If you feel strongly about it, safely merging the query strings params would be better.

**Comment by sballesteros** [comment]
Adding @halmos

**Comment by sballesteros** [comment]
I wonder if we should not discuss the UI a bit more here.
For context, the current UI was designed to be an overlay UI (to remain as consistent with what the extension provides) and here we are changing that (a bit). 
We decided to go with an overlay UI as we don't have control on the content where the shell is being injected in the extension case.

Here we could stick to an overlay UI and generalize the current shell so that the user can control the docking position (bottom, left or right). Note that we could also use different initial docking position based on the screen size if desired or let the user dock the shell where wanted.
Under that option, we would adapt the shell UI so the drag icon (allowing to resize the shell) can exists when the shell is docked to the left or right (so the shell would need some sort of border on the left or right).

**Comment by georgiamoon** [comment]
I agree about moving towards a dockable UI that the user can control. IMO that's the desired end state. I think though that this seems to make improvements in that direction, so I would say let's make incremental progress and go ahead with these improvements. We could then follow up with enhancing the control to allow the user to choose where and how the shell is docked.

**Comment by sballesteros** [comment]
the logic is different though so not sure it's really incremental

File: src/utils/preprints.js (hunk 5 above)
Comment (sajacy) [reply src/utils/preprints.js]: Yep - reverted!  Resolving!

**Comment by sajacy** [comment]
Addressed the PDF + debounce comments.

#### Re: layout

I'm open to suggestions on how to restructure, but I did try to take a low-touch approach to implementing the side-by-side view: the CSS is the main driver for partitioning into docked vs. overlay layouts (which it kinda already was, because of the `--wide/narrow-screen` rules - I mainly just adjusted the left/width props).

#### Re: extension
I didn't really look too much into the extension code - is the idea that the extension can inject the overlay on any tab, e.g. directly on the preprint source PDFs?  Probably that means you want to keep the changes isolated to the fallback / site, and we probably need to keep an untouched `shell.css` for the extension? (e.g. `extension-shell.css` without all the `--wide-screen` stuff)

**Comment by sballesteros** [comment]
yes exactly for the extension, it injects the shell to preprint pages.
Right now the changes would "break" the extension in the sense that on wide screens the shell would hide part of the content.  See the screenshot below for an example:
Screen Shot 2020-04-07 at 9 55 51 PM

re layout, I would like to get @halmos perspective (if he has time) and discuss a bit more before settling on an approach. 
I am concerned that the current approach could be annoying based on certain screen sizes (see screenshot below where IMHO the PDF is too small). 

Screen Shot 2020-04-07 at 10 01 52 PM

The overlay approach may be more consistent and more flexible (and I don't think that there are any technical blockers to make it happen). That being said we could find tune the media query etc.

**Comment by halmos** [comment]
My POV is that attention and care should be given to how the shell appears at different screen sizes. My understanding is that the right-side position is only for larger screens - is that correct? 

It seems to me that the side-by-side layout is only viable at very wide screens, but assuming that the size is large enough to accommodate a legible pdf rendering (and without it being obscured by the side-panel), then I don't see any issue with it. 

To @sballesteros point - in "overlay" mode, the bottom position has several usability benefits in that the text is still readable while the pdf is partially obscured. When overlayed on the side (ie the side-panel partially covers the right side of the pdf), that is not possible. So for the side-dock, I would discourage using an overlay and instead make it a side-by-side layout, and, again, only when the screen is wide enough to not reduce the size of the PDF so that accessibility concerns come into play. Unfortunately, for the extension,  that may require a more significant refactoring.

**Comment by sballesteros** [comment]
Thanks! good point on the fact that right / left overlay is less useful than bottom one when there is overlap.
One extra complexity for the non overlay version and the native PDF reader is that user can set the PDF to cover 100% of width:
Screen Shot 2020-04-08 at 7 28 19 AM
That could be a reason to still implement the right docked version as an overlay so that if user resize the PDF user can re-adjust / hide the shell

**Comment by sajacy** [comment]
Please review the branch that preserves the extension's overlay layout, (basically retains the old CSS), and let me know if I can add that commit here.

**Comment by sballesteros** [comment]
thanks! just left a question comment there. 
Otherwise (and unrelated) we should call `.cancel()` on the debounced function when we clean up.

File: src/components/shell.js [hunk 7]
```diff
@@ -94,13 +102,19 @@ export default function Shell({ children, defaultStatus = 'default' }) {
       }
     }
 
+    // ensure we render height as side-effect of page load
+    renderShellHeight();
+
+    const debouncedWindowResize = debounce(renderShellHeight, 500);
```
Comment (sballesteros) [inline src/components/shell.js:108]: need to call `debouncedWindowResize.cancel()` when we cleanup

Comment (sajacy) [reply src/components/shell.js:108]: good catch! added.

**Comment by georgiamoon** [comment]
@sajacy sorry to be late to follow up here, but I was just doing some testing on this locally and noticed that the mobile sizing is being a bit strange. I'm not 100% sure what's causing it, but if you could check on this, it seems like the PDF is zooming out so that it's really small.

Screenshot:
Screen Shot 2020-04-14 at 9 48 02 PM

Aside from this, I think this is a definitely improvement and so I'd say that from a UX perspective this LGTM.

**Comment by sballesteros** [comment]
FWIW I left a comment in https://github.com/PREreview/rapid-prereview/pull/88#issuecomment-611348705

**Comment by sajacy** [comment]
@georgiamoon The PDF width issue is fixed.  TIL: Responsive Design Mode.  Tested in all emulated devices on Firefox, Safari, Chrome on Mac.  Feel free to test further and merge at will.

**Comment by georgiamoon** [comment]
Thanks @sajacy !

**Pull request merged**